rpc/dns_monitor.go

```diff
@@ -0,0 +1,203 @@
+package rpc
+
+import (
+	"context"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+// DNSMonitor handles background DNS monitoring for worker gateways
+type DNSMonitor struct {
+	enabled       bool
+	checkInterval time.Duration
+	connectionStr string
+	ctx           context.Context
+	cancel        context.CancelFunc
+	stopComplete  chan struct{}
+}
+
+var (
+	// Global DNS monitor instance
+	dnsMonitor     *DNSMonitor
+	dnsMonitorLock sync.Mutex
+
+	// Prevent concurrent reconnections - global state independent of monitor lifecycle
+	reconnectionInProgress atomic.Bool
+
+	// Rate limiting state - global to survive monitor restarts during reconnection
+	lastReconnectTime  time.Time
+	lastReconnectMutex sync.Mutex
+)
+
+// StartDNSMonitor initializes and starts the background DNS monitor
+func StartDNSMonitor(enabled bool, checkInterval int, connectionString string) {
+	dnsMonitorLock.Lock()
+	defer dnsMonitorLock.Unlock()
+
+	// Stop existing monitor if running
+	if dnsMonitor != nil {
+		stopDNSMonitorInternal()
+	}
+
+	// Don't start if disabled
+	if !enabled {
+		Log.Debug("DNS monitor is disabled")
+		return
+	}
+
+	// Validate configuration
+	if connectionString == "" {
+		Log.Warning("DNS monitor enabled but connection string is empty, skipping monitor start")
+		return
+	}
+
+	// Set default interval if not specified or invalid
+	if checkInterval <= 0 {
+		checkInterval = 30 // Default to 30 seconds
+	}
```

Comment:
not only check that is greater than zero, but also set a minimum accepted value, for exmple if someone sets 1 second is currently valid, but it will lead with issues, because it will check every second and therefore we will get blocked in the DNS server